lib/buffer.js

```diff
@@ -24,82 +24,148 @@ function createPool() {
-    if (typeof encoding !== 'string' || encoding.length === 0)
-      encoding = 'utf8';
-    this.length = Buffer.byteLength(subject, encoding);
+    return new Buffer(arg);
+  }
 
-  // Handle Arrays, Buffers, Uint8Arrays or JSON.
-  } else if (subject !== null && typeof subject === 'object') {
-    if (subject.type === 'Buffer' && Array.isArray(subject.data))
-      subject = subject.data;
-    this.length = +subject.length;
+  this.length = 0;
+  this.parent = undefined;
 
-  } else {
-    throw new TypeError('must start with number, buffer, array or string');
+  // Common case.
+  if (typeof(arg) === 'number') {
+    fromNumber(this, arg);
+    return;
   }
 
-  if (this.length > kMaxLength) {
-    throw new RangeError('Attempt to allocate Buffer larger than maximum ' +
-                         'size: 0x' + kMaxLength.toString(16) + ' bytes');
+  // Slightly less common case.
+  if (typeof(arg) === 'string') {
+    fromString(this, arg, arguments.length > 1 ? arguments[1] : 'utf8');
+    return;
   }
 
-  if (this.length < 0)
-    this.length = 0;
-  else
-    this.length >>>= 0;  // Coerce to uint32.
+  // Unusual.
+  fromObject(this, arg);
+}
 
-  this.parent = undefined;
-  if (this.length <= (Buffer.poolSize >>> 1) && this.length > 0) {
-    if (this.length > poolSize - poolOffset)
-      createPool();
-    this.parent = sliceOnto(allocPool,
-                            this,
-                            poolOffset,
-                            poolOffset + this.length);
-    poolOffset += this.length;
-  } else {
-    alloc(this, this.length);
-  }
+function fromNumber(that, length) {
+  allocate(that, length < 0 ? 0 : checked(length) >>> 0);
+}
 
-  if (typeof subject === 'number') {
-    return;
+function fromString(that, string, encoding) {
+  if (typeof(encoding) !== 'string' || encoding === '')
+    encoding = 'utf8';
+
+  // Assumption: byteLength() return value is always < kMaxLength.
+  var length = byteLength(string, encoding) >>> 0;
+  allocate(that, length);
+
+  var actual = that.write(string, encoding) >>> 0;
+  if (actual !== length) {
+    // Fix up for truncated base64 input.  Don't bother returning
+    // the unused two or three bytes to the pool.
+    that.length = actual;
+    truncate(that, actual);
   }
+}
 
-  if (typeof subject === 'string') {
-    // In the case of base64 it's possible that the size of the buffer
-    // allocated was slightly too large. In this case we need to rewrite
-    // the length to the actual length written.
-    var len = this.write(subject, encoding);
-    // Buffer was truncated after decode, realloc internal ExternalArray
-    if (len !== this.length) {
-      var prevLen = this.length;
-      this.length = len;
-      truncate(this, this.length);
-      // Only need to readjust the poolOffset if the allocation is a slice.
-      if (this.parent != undefined)
-        poolOffset -= (prevLen - len);
-    }
+function fromObject(that, object) {
+  if (object instanceof Buffer)
```

Comment:
once it reaches C++ buffers and smalloc allocations are essentially treated the same.